packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -750,8 +752,8 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
   }
 
   // Called when _textEditingController changes.
-  void _onChangedField() {
-    final Iterable<T> options = widget.optionsBuilder(
+  Future<void> _onChangedField() async {
+    final Iterable<T> options = await widget.optionsBuilder(
```

Comment:
This looks great, is it really this easy?  I want to make sure these cases are easy too:

  1. What if the async call fails?  Would it be easy to show an error state, or another option, to show the previous options instead?
  2. Would it be easy to show a loading state?
  2. What if an async call finishes out of order?  Similarly, what if the developer wants to cancel/ignore a pending call when the input changes?
  3. What if you want to debounce the async calls?

Maybe these are the responsibility of the developer and not the framework...  I want to make sure they're easy to handle though.  What do you think?

Reply:
3. The "out of order" case is interesting. What would be the expected result? I guess the developer should be in charge of canceling ongoing operations, but I'm not sure what would be a nice way to indicate it to the framework that a previous call should be ignored.

Reply:
Well the expected result from the user's perspective would definitely be that the options reflect the current state of the input field.  I think the developer could either cancel old requests, or they could compare the state of the input field at the time a request was sent and when it returns.

I think I'm pretty convinced that all of my points above are the responsibility of the developer, and they're all solvable with this simple FutureOr setup.  It's going to be tempting for developers to just call their API directly in optionsBuilder, but really they should have some sort of async abstraction over it that handles all of the points above.  Maybe if I get some time I can write some examples or something.